### PR TITLE
Custom region masks

### PR DESCRIPTION
**Pull request by lewisjared**
# Pull request

Please confirm that this pull request has done the following:

- [x] Tests added
- [x] Documentation added (where applicable)
- [x] Description in ``CHANGELOG.rst`` added

This PR refactors the masking of regions in the `masks` module and adds the "World|North Atlantic|Ocean" region which isn't currently used. I'll add functionality to specify regions via the CLI in a separate PR.

### CODE REVIEW

File: docs/source/usage.rst [hunk 1]
```diff
@@ -15,3 +15,5 @@ In a masked array, the convention is that a value of ``True`` in a mask indicate
 Hence a ``land_mask`` should be ``False`` where the associated data is land and ``True`` where the associated data is ocean.
 This can be confusing as then a land and northern hemisphere mask have to be combined with 'or' logic rather than 'and' logic i.e. ``land_nh_mask = land_mask or nh_mask``, not ``land_nh_mask = land_mask and nh_mask`` as one might intuitively expect.
 (One way to think about it is that we want to mask where we're not on land or where we're not in the northern hemisphere, leaving only the regions where we're both on land and in the northern hemisphere).
+
+See the :doc:`masks` documentation for examples of what masks are available.
```
Comment (znicholls) [inline docs/source/usage.rst]:
```suggestion
See the :doc:`masks` documentation for examples of the available masks.
```

File: src/netcdf_scm/masks.py [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+
+This is a mapping between mask name and a MaskFunc. These MaskFunc's are called by an instance of ``CubeMasker``.
+
+A MaskFunc is a function which takes a ScmCube, CubeMasker and any additional keyword arguments. The function should return a numpy
+array of boolean's with the same dimensionality as the ScmCube. Where True values are returned, data will be masked out (excluded)
+from any calculations. The "World|Northern Hemisphere|Land" mask should be True everywhere except for land cells in the Northern 
+Hemisphere.
+
+These MaskFunc's can be composed together to create more complex functionality. For example 
+`or_masks(get_area_mask(0, -80, 65, 0), "World|Ocean")` will return a mask consisting of an or between a subsetted area and an ocean mask.
```
Comment (znicholls) [inline src/netcdf_scm/masks.py]:
```suggestion
`or_masks(get_area_mask(0, -80, 65, 0), "World|Ocean")` will return the result of an 'or' operation between a subsetted area and an ocean mask.
```

File: src/netcdf_scm/masks.py [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+
+        The result only contains valid masks. Any invalid masks are dropped.
+        """
+        masks = {}
+        for name in mask_names:
+            try:
+                mask = self.get_mask(name)
+                masks[name] = mask
+            except InvalidMask:
+                pass
```
Comment (znicholls) [inline src/netcdf_scm/masks.py]: do we want an error/warning here explaining what happened

File: src/netcdf_scm/utils.py [hunk 4]
```diff
@@ -177,6 +178,32 @@ def unify_lat_lon(cubes, rtol=10 ** -10):
         cube.add_dim_coord(cubes[0].coords("longitude")[0], lon_dim_no)
 
 
+def broadcast_onto_lat_lon_grid(cube, array_in):
+    """Broadcast an array onto the latitude-longitude grid of ``cube``.
+
+    Here, broadcasting means taking the array and 'duplicating' it so that it
+    has the same number of dimensions as the cube's underlying data. For example,
+    if our cube has a time dimension of length 3, a latitude dimension of length 4
+    and a longitude dimension of length 2 then if we are given in a 4x2 array, we
```
Comment (znicholls) [inline src/netcdf_scm/utils.py]:
```suggestion
    and a longitude dimension of length 2 then if we are ``array_in`` is a 4x2 array, we
```

**Comment by lewisjared** [comment]
@znicholls Any idea how to stop codacy flagging asserts in tests?

File: src/netcdf_scm/masks.py (hunk 3 above)
Comment (lewisjared) [reply src/netcdf_scm/masks.py]: The exception message won't be all that helpful here as it won't always be due to the root cause of it failing in the case of cascading failures.
I.e missing land mask > can't make ocean mask > can't make NH|Ocean mask

**Comment by znicholls** [comment]
> Any idea how to stop codacy flagging asserts in tests?

Yes but for now I don't care and will just merge anyway haha

File: src/netcdf_scm/masks.py [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+def get_area_mask(lower_lat, left_lon, upper_lat, right_lon):
+    """
+    Mask a subset of the globe
+
+    Circular coordinates (longitude) can cross the 0E.
+
+    Parameters
+    ----------
+    lower_lat: int or float
+    left_lon: int or float
```
Comment (znicholls) [inline src/netcdf_scm/masks.py]:
```suggestion
    left_lon: int or float of degrees East
```

File: src/netcdf_scm/masks.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+    upper_lat: int or float
+    right_lon: int or float
```
Comment (znicholls) [inline src/netcdf_scm/masks.py]:
```suggestion
    right_lon: int or float of degrees East
```

File: src/netcdf_scm/masks.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+    # to False, do all our operations with AND logic, then flip everything
+    # back).
+    mask_nh = ~np.outer(~mask_nh_lat, ~mask_all_lon)
+
+    return broadcast_onto_lat_lon_grid(cube, mask_nh)
+
+
+def get_area_mask(lower_lat, left_lon, upper_lat, right_lon):
+    """
+    Mask a subset of the globe
```
Comment (znicholls) [inline src/netcdf_scm/masks.py]:
```suggestion
    Mask a subset of the globe using latitudes and longitudes in degrees East
```

File: src/netcdf_scm/masks.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+    "World|Southern Hemisphere|Land": or_masks(
+        "World|Southern Hemisphere", "World|Land"
+    ),
+    "World|Northern Hemisphere|Ocean": or_masks(
+        "World|Northern Hemisphere", "World|Ocean"
+    ),
+    "World|Southern Hemisphere|Ocean": or_masks(
+        "World|Southern Hemisphere", "World|Ocean"
+    ),
+    "World|North Atlantic|Ocean": or_masks(get_area_mask(0, -80, 65, 0), "World|Ocean"),
```
Comment (znicholls) [inline src/netcdf_scm/masks.py]:
```suggestion
    "World|North Atlantic Ocean": or_masks(get_area_mask(0, -80, 65, 0), "World|Ocean"),
```

Is 'North Atlantic' not ocean by definition?

File: src/netcdf_scm/masks.py (hunk 3 above)
Comment (znicholls) [reply src/netcdf_scm/masks.py]: can we make it a warning, "Failed to get mask for 'NH|Ocean'" so users at least know? This is a silent fail which I'm uncomfortable with.

**Pull request merged**